main/src/ca/uwaterloo/flix/language/ast/Ast.scala

```diff
@@ -465,4 +465,43 @@ object Ast {
     */
   case class Derivation(clazz: Symbol.ClassSym, loc: SourceLocation)
 
+  /**
+    * Represents the way a variable is bound.
+    */
+  sealed trait BoundBy
+
+  object BoundBy {
+
+    /**
+      * Represents a variable that is bound by a formal parameter.
+      */
+    case object Param extends BoundBy
+
+    /**
+      * Represents a variable that is bound by a let.
+      */
+    case object Let extends BoundBy
+
+    /**
+      * Represents a variable that is bound by a pattern.
+      */
+    case object Pattern extends BoundBy
+
+    /**
+      * Represents a variable that is bound by a select.
+      */
+    case object Select extends BoundBy
+
+    /**
+      * Represents a variable that is bound by a constraint.
+      */
+    case object Constraint extends BoundBy
+
+    /**
+      * Represents some other unknown binding.
+      */
+    case object Unclassified extends BoundBy
```

Comment:
I'd prefer to avoid this if at all possible. See other comment.

Reply:
How detailed would you want this then?

Reply:
The only other case I see here is the "CatchRule" binding. But I guess we might have a lot more if it becomes a required field?